triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/AuthServiceClient.scala

```diff
@@ -29,20 +57,116 @@ class AuthServiceClient(authServiceBaseUri: Uri)(
+    val req = HttpRequest(
       method = HttpMethods.POST,
-      uri = authorizeUri,
+      uri,
       headers = List(Authorization(BasicHttpCredentials(username, password)))
     )
-    for {
-      authResponse <- http.singleRequest(request)
-      authServiceToken <- Unmarshal(authResponse).to[AuthServiceToken]
-    } yield authServiceToken
+    http.singleRequest(req).flatMap(Unmarshal(_).to[AuthServiceToken])
   }
 
+  def requestServiceAccount(
+      authServiceToken: AuthServiceToken,
+      ledgerId: String): Future[Boolean] = {
+    val uri = authServiceBaseUri.withPath(saSecure./("request")./(ledgerId))
+    val authHeader = Authorization(OAuth2BearerToken(authServiceToken.token))
+    val nonce = Nonce(UUID.randomUUID.toString)
+    val entity = HttpEntity(ContentTypes.`application/json`, nonce.toJson.toString)
+    val req = HttpRequest(
+      method = HttpMethods.POST,
+      uri,
+      headers = List(authHeader),
+      entity
+    )
+    http.singleRequest(req).map(_.status.isSuccess)
+  }
+
+  def listServiceAccounts(authServiceToken: AuthServiceToken): Future[ServiceAccountList] = {
+    val uri = authServiceBaseUri.withPath(saSecure)
+    val authHeader = Authorization(OAuth2BearerToken(authServiceToken.token))
+    val req = HttpRequest(
+      method = HttpMethods.GET,
+      uri,
+      headers = List(authHeader),
+    )
+    http.singleRequest(req).flatMap(Unmarshal(_).to[ServiceAccountList])
+  }
+
+  def getServiceAccount(authServiceToken: AuthServiceToken): Future[Option[ServiceAccount]] =
+    try {
+      RetryStrategy.constant(attempts = 3, waitTime = 4.seconds) { (_, _) =>
+        for {
+          ServiceAccountList(sa :: _) <- listServiceAccounts(authServiceToken)
+        } yield Some(sa)
+      }
+    } catch {
+      case e: Throwable => Future(None)
+    }
+
+  def requestCredential(
+      authServiceToken: AuthServiceToken,
+      serviceAccountId: String): Future[Boolean] = {
+    val uri = authServiceBaseUri.withPath(saSecure./(serviceAccountId)./("credRequest"))
+    val authHeader = Authorization(OAuth2BearerToken(authServiceToken.token))
+    val req = HttpRequest(
+      method = HttpMethods.POST,
+      uri,
+      headers = List(authHeader),
+    )
+    http.singleRequest(req).map(_.status.isSuccess)
+  }
+
+  def getNewCredentialId(
+      authServiceToken: AuthServiceToken,
+      serviceAccountId: String): Future[Option[CredentialId]] =
+    getServiceAccount(authServiceToken) flatMap {
+      case None => Future(None)
+      case Some(sa) =>
+        val numCreds = sa.creds.length
+        requestCredential(authServiceToken, sa.serviceAccount) flatMap { reqSuccess =>
+          if (!reqSuccess) Future(None)
+          else
+            try {
+              RetryStrategy.constant(attempts = 3, waitTime = 4.seconds) { (_, _) =>
+                for {
+                  ServiceAccountList(sa :: _) <- listServiceAccounts(authServiceToken)
```

Comment:
Is there a guarantee on the order here?

Reply:
It gets added to the front of the list in the implementation, will check that it's part of the API.